src/dispatch.c

```diff
@@ -82,3 +83,452 @@ natsStatus natsSub_enqueueUserMessage(natsSubscription *sub, natsMsg *msg)
     return NATS_OK;
 }
 
+static inline void _removeFromDispatchQueue(natsDispatcher *d, natsMsg *msg)
+{
+    d->queue.head = msg->next;
+    if (d->queue.tail == msg)
+        d->queue.tail = NULL;
+    msg->next = NULL;
+}
+
+// Returns fetch status
+static inline natsStatus _preProcessUserMessage(
+    natsSubscription *sub, jsSub *jsi, jsFetch *fetch, natsMsg *msg,
+    bool *userMsg, bool *overLimit, bool *lastMessageInSub, bool *lastMessageInFetch, char **fcReply)
+{
+    natsStatus fetchStatus = NATS_OK;
+    *userMsg = true;
+
+    // Is this a real message? If so, account for having processed it.
+    bool isRealMessage = (msg->subject[0] != '\0');
+    if (isRealMessage)
+    {
+        sub->ownDispatcher.queue.msgs--;
+        sub->ownDispatcher.queue.bytes -= natsMsg_dataAndHdrLen(msg);
+    }
+
+    // Fetch-specific handling of synthetic and header-only messages
+    if (fetch != NULL)
+        fetchStatus = js_checkFetchedMsg(sub, msg, true, userMsg);
```

Comment:
This function says that the lock must be held, but looks like the caller of the current function is not in some cases (in cats_dispatchThreadPool).

Reply:
Thanks for the catch, restructured so that no locking required.